snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,300 @@
+            """
+            Applies the appropriate optimization on a single arrayJoin expression.
+            """
+            match = arrayjoin_pattern.match(expr)
+            if match is None:
+                return expr
+
+            if arrayjoins_in_query == {"tags.key", "tags.value"}:
+                # Both tags_key and tags_value expressions present int the query.
+                # do the arrayJoin on key-value pairs instead of independent
+                # arrayjoin for keys and values.
+                array_index = (
+                    LiteralExpr(None, 1)
+                    if match.string("col") == "tags.key"
+                    else LiteralExpr(None, 2)
+                )
+
+                if not filtered_tags:
+                    return _unfiltered_tag_pairs(expr.alias, array_index)
+                else:
+                    return _filtered_tag_pairs(expr.alias, filtered_tags, array_index)
+
+            elif filtered_tags:
+                # Only one between tags_key and tags_value is present, and
+                # it is tags_key since we found filtered tag keys.
+                return _filtered_tag_keys(expr.alias, filtered_tags)
+            else:
+                # No viable optimization
+                return expr
+
+        query.transform_expressions(replace_expression)
+
+
+def _unfiltered_tag_pairs(alias: Optional[str], array_index: LiteralExpr) -> Expression:
+    # (arrayJoin(
+    #   arrayMap((x,y) -> [x,y], tags.key, tags.value)
+    #  as all_tags)[1]
+    return arrayElement(
+        alias,
+        array_join(
+            "all_tags",
+            zip_columns(
+                ColumnExpr(None, None, "tags.key"),
+                ColumnExpr(None, None, "tags.value"),
+            ),
+        ),
+        array_index,
+    )
+
+
+def _filtered_tag_pairs(
+    alias: Optional[str],
+    filtered_tags: Sequence[LiteralExpr],
+    array_index: LiteralExpr,
+) -> Expression:
+    # (arrayJoin(arrayFilter(
+    #       pair -> arrayElement(pair, 1) IN (tags),
+    #       arrayMap((x,y) -> [x,y], tags.key, tags.value)
+    #  )) as all_tags)[1]
+    return arrayElement(
+        alias,
+        array_join(
+            "all_tags",
+            filter_key_values(
+                zip_columns(
+                    ColumnExpr(None, None, "tags.key"),
+                    ColumnExpr(None, None, "tags.value"),
+                ),
+                filtered_tags,
+            ),
+        ),
+        array_index,
+    )
+
+
+def _filtered_tag_keys(
+    alias: Optional[str], filtered_tags: Sequence[LiteralExpr]
+) -> Expression:
+    # arrayJoin(arrayFilter(
+    #   tag -> tag IN (tags),
+    #   tags.key
+    # ))
+    return array_join(
+        alias, filter_keys(ColumnExpr(None, None, "tags.key"), filtered_tags),
+    )
+
+
+def array_join(alias: Optional[str], content: Expression) -> Expression:
+    return FunctionCallExpr(alias, "arrayJoin", (content,))
```

Comment:
I generally did not put them there at first, because they are used only in one place now and I did not really want to encourage people to use arrayJoin, but very weak opinion.

Reply:
I moved arrayJoin